server/src/main/java/io/crate/auth/JWTAuthenticationMethod.java

```diff
@@ -0,0 +1,164 @@
+
+        Role user = roles.findUser(credentials.username());
+        if (user == null) {
+            throw new RuntimeException("jwt authentication failed for user \"" + credentials.username() + "\"");
+        }
+        assert user.jwtProperties() != null : "Resolved user must have jwt properties";
+
+        try {
+            JwkProvider jwkProvider = jwkProvider(credentials);
+            RSAKeyProvider keyProvider = new RSAKeyProvider() {
+                @Override
+                public RSAPublicKey getPublicKeyById(String kid) {
+                    try {
+                        return (RSAPublicKey) jwkProvider.get(kid).getPublicKey();
+                    } catch (JwkException e) {
+                        throw new RuntimeException("jwt authentication failed for user \"" + credentials.username() + "\"", e);
+                    }
+                }
+
+                @Override
+                public RSAPrivateKey getPrivateKey() {
+                    return null;
+                }
+
+                @Override
+                public String getPrivateKeyId() {
+                    return null;
+                }
+            };
+
+            Algorithm algorithm;
+            switch (credentials.decodedToken().getAlgorithm()) {
+                case "RS256" :
+                    algorithm = Algorithm.RSA256(keyProvider);
+                    break;
+                case "RS384" :
+                    algorithm = Algorithm.RSA384(keyProvider);
+                    break;
+                case "RS512" :
+                    algorithm = Algorithm.RSA512(keyProvider);
+                    break;
+                default:
+                    // Not supposed to happen, we validate token right after extracting from header and decoding.
+                    throw new RuntimeException("jwt authentication failed for user \"" + credentials.username() + "\", unsupported algorithm");
+            }
+
+            // Expiration date is checked by default(if provided in token)
+            JWTVerifier.BaseVerification verification = (JWTVerifier.BaseVerification) JWT.require(algorithm)
+                // withers below are not needed for payload signature check.
+                // It's an extra step on top of signature verification to double check that user metadata matches token payload.
+                .withIssuer(user.jwtProperties().iss())
+                .withClaim("username", user.jwtProperties().username());
+
+            JWTVerifier verifier = verification.build(clock());
+
+            verifier.verify(credentials.decodedToken());
+
+        } catch (Exception e) {
+            throw new RuntimeException("jwt authentication failed for user \"" + credentials.username() + "\"", e);
+        }
+
+        return user;
+    }
+
+    public Clock clock() {
+        return Clock.systemUTC();
+    }
+
+    public JwkProvider jwkProvider(Credentials credentials) {
+        URL jwkUrl;
```

Comment:
Separate method for clock/jwk provider only for the sake of test simplicity, see `TestJWTAuthenticationMethod`